### DISCUSSION TITLE
Model issues in producing variable graphs

### DISCUSSION THREAD

**Discussion by rokapary** [JenniNiku/gllvm#188] [Q&A] [answered]
Hello,

I've been tasked with exploring species correlations and the effect of environmental variables on them, with the GLLVM approach being chosen as a good fit for the goals. As this is my first time using a GLLVM, I was given the tutorial held by this group in 2020 (and your other guides) as a sort of manual for how to achieve the desired results. However, I have run into some issues.

The data I am using is:
speciestable.sub – a 500 row subset of 26 species presence/absence data (1/0)
climatetable.sub – a 500 row data table containing 19 climatic variables, which have been standardized to a mean of 0 and SD of 1 using the scale() function.

When I fit the model as shown:
fit_baseSP1 <- gllvm(speciestable.sub, num.lv = 2, family = binomial(link="probit"), sd.errors = FALSE)

I receive the error message: “#In gllvm(speciestable.sub, num.lv = 2, family = binomial(link = "probit"),  : There are responses full of zeros.”

I saw that someone else here had a similar problem and the explanation given is that the warning is thrown when there are sites full of zero. However, all sites that contain only zeroes have already been excluded from this data set. Even when experimentally excluding sites with only 1 species present the warning persists.

This becomes a bigger issue when I try to use the generated GLLVM for the process of exploring trends between the env. variable and latent variables. The graphs I receive look like this:

![En_Cov](https://github.com/JenniNiku/gllvm/assets/75184957/bc6ef823-7f3d-4195-ada3-2cc7ed5b179c)

The code used being:
for (i in 1:length(colnames(climatetable.sub))) {
  covariate <- climatetable.sub[,i]
  rbPal <- colorRampPalette(c('red', 'blue'))
  Colorsph <- rbPal(20)[as.numeric(cut(covariate, breaks = 20))]
  breaks <- seq(min(covariate), max(covariate), length.out = 30) 
  ordiplot(fit_baseSP1, main = paste0("Ordination of sites, color:",colnames(climatetable.sub)[i]),
           symbols = TRUE, s.colors = Colorsph, xlim = c(-1.2,1.2), ylim = (c(-1.2, 1.2)))
}

As it stands I’m not sure where the problem lies. Any advice would be greatly appreciated. 

Many thanks!

**Comment by BertvanderVeen** [comment]
Hello! The warning

`there are responses full of zeros.`

means that there are <b>species</b> that only contain zeros. These should be omitted.

The ordination that you are showing is telling you that there is only one LV necessary, the second LV is zero. Note, that 19 environmental covariates is already quite a lot, and that you need at least 19+1+num.lv observations for every species to succesfully estimate all of these parameters. Do you have that much information for all the species?

Note that if you are setting the model up as you are, the covariate effects are <b>removed</b> from the ordination, and there should be no relationship with the ordination and the covariates. If you are interested in relating the covariates to the ordination, have a look at [this vignette](https://jenniniku.github.io/gllvm/articles/vignette6.html) how to do that more appropriately with constrained ordination instead.

**Reply by rokapary** [reply]
Hello, 

Many thanks for your feedback! Removing the "0" species and changing the level number really did help. 

As for your further points, it does seem that some of the species don't have the neccesary number of observations even at bigger data subsets. In which case, I suppose I should cut down the variable number before continuing? Many of them are highly correlated, so that may not be too difficult.

My orginal goal was to recreate ths vignette: https://jenniniku.github.io/gllvm/articles/vignette4.html to see how the covariates affect the correlation matrix. But if as it says the vignette you sent and the number of species and covariates is too large, then the constrained ordination is the better alternative, correct?

Again, many thanks for your help! It is highly appreciated.

**Reply by BertvanderVeen** [reply, marked as answer]
Yes! If you have many covariates and not so much information, *constrained*, *concurrent*, or partial constrained ordination (num.RR +num.lv) can help. You could try fitting first an unconstrained ordination, look at the residual correlation plot, then change to a partial constrained ordination and do the same again, which still recreates the vignette in a similar fashion!
